server/server.go

```diff
@@ -6,53 +6,299 @@ package server
 
 import (
 	"context"
+	"errors"
 	"fmt"
-
-	pbts "google.golang.org/protobuf/types/known/timestamppb"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
 
 	"github.com/elastic/elastic-agent-libs/logp"
 	pb "github.com/elastic/elastic-agent-shipper-client/pkg/proto"
 	"github.com/elastic/elastic-agent-shipper-client/pkg/proto/messages"
 	"github.com/elastic/elastic-agent-shipper/queue"
+
+	"github.com/gofrs/uuid"
 )
 
+// ShipperServer contains all the gRPC operations for the shipper endpoints.
+type ShipperServer interface {
+	pb.ProducerServer
+	io.Closer
+}
+
 type shipperServer struct {
 	logger *logp.Logger
+	queue  *queue.Queue
+	cfg    ShipperServerConfig
+
+	uuid           string
+	acceptedIndex  int64
+	persistedIndex int64
 
-	queue *queue.Queue
+	polling       polling
+	notifications notifications
+
+	close *sync.Once
 
 	pb.UnimplementedProducerServer
 }
 
-// PublishEvents is the server implementation of the gRPC PublishEvents call
-func (serv shipperServer) PublishEvents(_ context.Context, req *messages.PublishRequest) (*messages.PublishReply, error) {
-	results := []*messages.EventResult{}
-	for _, evt := range req.Events {
-		serv.logger.Infof("Got event %s: %#v", evt.EventId, evt.Fields.Data)
-		err := serv.queue.Publish(evt)
-		if err != nil {
-			// If we couldn't accept any events, return the error directly. Otherwise,
-			// just return success on however many events we were able to handle.
-			if len(results) == 0 {
-				return nil, err
+type polling struct {
+	ctx     context.Context
+	stop    func()
+	stopped chan struct{}
+}
+
+// NewShipperServer creates a new server instance for handling gRPC endpoints.
+func NewShipperServer(q *queue.Queue, cfg ShipperServerConfig) (ShipperServer, error) {
+	if q == nil {
+		return nil, errors.New("queue cannot be nil")
+	}
+
+	err := cfg.Validate()
+	if err != nil {
+		return nil, fmt.Errorf("invalid configuration: %w", err)
+	}
+
+	id, err := uuid.NewV4()
+	if err != nil {
+		return nil, err
+	}
+
+	s := shipperServer{
+		uuid:   id.String(),
+		logger: logp.NewLogger("shipper-server"),
+		queue:  q,
+		cfg:    cfg,
+		polling: polling{
+			stopped: make(chan struct{}),
+		},
+		notifications: notifications{
+			subscribers: make(map[uuid.UUID]chan change),
+			mutex:       &sync.Mutex{},
+		},
+		close: &sync.Once{},
+	}
+
+	s.polling.ctx, s.polling.stop = context.WithCancel(context.Background())
+	s.startPolling()
+
+	return &s, nil
+}
+
+// GetAcceptedIndex atomically reads the accepted index
+func (serv *shipperServer) GetAcceptedIndex() int64 {
+	return atomic.LoadInt64(&serv.acceptedIndex)
+}
+
+// GetPersistedIndex atomically reads the persisted index
+func (serv *shipperServer) GetPersistedIndex() int64 {
+	return atomic.LoadInt64(&serv.persistedIndex)
```

Comment:
I made another PR to expose a placeholder on the queue for the pending work at https://github.com/elastic/elastic-agent-shipper/pull/77. These are calls on `serv.queue` that can be used here and then filled in once the libbeat work is done.